# Bio-sequence Alphabet Expansion

This repository is ground zero for my project to expand the bio-sequence
alphabet to allow a wider range of chemical entities as part of bio-sequences,
i.e. post-translational modifications of proteins and DNA modifications.

The general idea is summarized in the following graphic.

![Project-Central-Idea.jpg A visual and textual explanation of the project ideas and goals.](/files/Project-Central-Idea.jpg)


# Contents

```
.
├── Fabian\ Egli\ fabian.egli@gmx.ch\ (0x63E17ADB7D4734B4)\ pub.asc
├── Glossary.md
├── Introduction.md
├── Project-Log.md
├── Proposal-BioFont.md
├── Proposal-UTF-8.md
├── Python.md
├── README.md
├── files
│   └── Project-Central-Idea.jpg
└── tests
    └── Python-Performance-index-access-speed.md
```


# Community Input

This is a great endeavor and requires input from many stakeholders. These
include all people who generate, store and work with bio-sequence information.

The input/feedback/proposals from anyone involved with bio-sequence data will be
considered. There are multiple ways to get your voice heard:

* *Become a Collaborator*: Become a part of the team shaping the future of
bio-sequence encoding. Write me an [email][1].
* Open an Issue in this repository.


# License

Copyright © 2019 by Fabian Egli.
All Rights Reserved.

This repository and all its contents are copyrighted to the fullest extent the
law permits. All contents, including, but not limited to, original ideas,
considerations on performance impacts of encoding changes, performance tests,
encoding proposals and font proposals may not be used for any purpose. You can
not use any content to acquire grants from any funding agency, except with
prior explicit written permission from me - [Fabia Emanuel Egli][1].


### Why this Restrictive License

I need money to life - and this is my project. Don't steal it. Once I get the
funding, I will release this repository and its contents under a permissive
license to benefit the scientific community. After all, this is the whole idea
of this project.


# Project milestones

1. Acquire funding to for this project.
2. Realize the project.
3. Release the project under the Attribution-ShareAlike 4.0 International [(CC
  BY-SA 4.0)](https://creativecommons.org/licenses/by-sa/4.0/)


# IUBMB Standard Definition

Procedure for publication:
https://iupac.org/what-we-do/recommendations/procedure-for-publication/

Guidelines for drafting reports:
https://iupac.org/what-we-do/recommendations/guidelines-for-drafting-reports/

Example definition:
https://www.degruyter.com/view/j/pac.2017.89.issue-11/pac-2016-0502/pac-2016-0502.xml

# Twitter handles of important entities

* [IUBMB](https://twitter.com/iubmb)
* [IUPAC](https://twitter.com/iupac)
* [SIB](https://twitter.com/ISBSIB)
* [UniProt](https://twitter.com/uniprot)
* [EMBL-EBI](https://twitter.com/emblebi)
* [Ensembl](https://twitter.com/ensembl)
* [The Unicode Consortium](https://twitter.com/unicode)


# Link Collection

http://dbptm.mbc.nctu.edu.tw/

IUPAC nucleic acids (15 single letter codes)
[https://www.ncbi.nlm.nih.gov/pmc/articles/PMC341218/](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC341218/)
[https://en.wikipedia.org/wiki/Nucleotide#cite_note-iupac-19](https://en.wikipedia.org/wiki/Nucleotide#cite_note-iupac-19)

### IUPAC amino acids (26 single letter codes)

IUPAC-IUB Joint Commission on Biochemical Nomenclature (JCBN), [https://link.springer.com/chapter/10.1007%2F978-94-009-7225-4_43](https://link.springer.com/chapter/10.1007%2F978-94-009-7225-4_43)


# Literature

* General overview on protein modifications [https://www.uniprot.org/help/mod_res](https://www.uniprot.org/help/mod_res)

### Unicode, ISO 10646 and UTF-8 encoding

* [https://en.wikipedia.org/wiki/UTF-8](https://en.wikipedia.org/wiki/UTF-8)
* Original UTF-8 paper "Hello World or Καλημέρα κόσμε or こんにちは 世界" by Rob Pike Ken Thompson, [http://doc.cat-v.org/plan_9/4th_edition/papers/utf](http://doc.cat-v.org/plan_9/4th_edition/papers/utf)
* The Unicode Standard v. 11.0.0 [https://www.unicode.org/versions/Unicode11.0.0/](https://www.unicode.org/versions/Unicode11.0.0/)

### Font & Glyph Design

* Taxonomy-Based Glyph Design—with a Case Study on Visualizing Workflows of Biological Experiments [10.1109/TVCG.2012.271](https://doi.org/10.1109/TVCG.2012.271)

## PTMs

### Glycosylation

* Library of named carbohydrates [http://csdb.glycoscience.ru/database/core/library.php#](http://csdb.glycoscience.ru/database/core/library.php#)
* DrawGlycan-SNFG [http://www.virtualglycome.org/DrawGlycan/](http://www.virtualglycome.org/DrawGlycan/)
* Carbohydrate Structure Database [http://csdb.glycoscience.ru/database/](http://csdb.glycoscience.ru/database/)
* Glynsight [https://glycoproteome.expasy.org/glynsight/](https://glycoproteome.expasy.org/glynsight/)
* ExPASy search for glycosylation [https://www.expasy.org/resources/search/keywords:glycosylation](https://www.expasy.org/resources/search/keywords:glycosylation)
* N-linked glycosylation (Wikipedia) [https://en.wikipedia.org/wiki/N-linked_glycosylation](https://en.wikipedia.org/wiki/N-linked_glycosylation)

### Phosphorylation

* Phospho.ELM database [http://phospho.elm.eu.org/](http://phospho.elm.eu.org/)
* ViralPhos [http://csb.cse.yzu.edu.tw/ViralPhos/](http://csb.cse.yzu.edu.tw/ViralPhos/)
* dbPAF [http://dbpaf.biocuckoo.org/index.php](http://dbpaf.biocuckoo.org/index.php)

### Others

* UniProt protein modifications [https://www.uniprot.org/help/mod_res](https://www.uniprot.org/help/mod_res)

# DNA and DNA Modifications

* An extended IUPAC nomenclature code for polymorphic nucleic acids, [https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btq098](https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btq098)

* Database or DNA Modifications [https://dnamod.hoffmanlab.org/](https://dnamod.hoffmanlab.org/)


# Contact

You can contact me by [email][1].

You can encrypt your messages to me with the following public key in this repository `Fabian Egli fabian.egli@gmx.ch (0x63E17ADB7D4734B4) pub.asc`.

[1]: mailto:fabian.egli@gmx.ch
